src/main.rs

```diff
@@ -1,3 +1,829 @@
-fn main() {
-    println!("Hello, world!");
+use std::collections::BTreeMap;
+use std::error::Error;
+use std::fmt;
+use std::io::{BufReader, Read};
+use std::ops::Deref;
+use std::path::Path;
+use std::{fs::File, io::Write, panic, path::PathBuf};
+
+use cargo_metadata::{Metadata, Package};
+use clap::{ArgEnum, CommandFactory, Parser, Subcommand};
+use log::{error, info, trace};
+use serde::de::Visitor;
+use serde::{de, de::Deserialize, ser::Serialize};
+use serde::{Deserializer, Serializer};
+use simplelog::{
+    ColorChoice, ConfigBuilder, Level, LevelFilter, TermLogger, TerminalMode, WriteLogger,
+};
+
+type Version = cargo_metadata::Version;
+
+#[derive(Parser)]
+#[clap(version, about, long_about = None)]
+#[clap(propagate_version = true)]
+struct Cli {
+    /// Subcommands ("no subcommand" is its own subcommand)
+    #[clap(subcommand)]
+    command: Option<Commands>,
+
+    // Cargo flags we support and forward to e.g. 'cargo metadata'
+    #[clap(flatten)]
+    manifest: clap_cargo::Manifest,
+    #[clap(flatten)]
+    workspace: clap_cargo::Workspace,
+    #[clap(flatten)]
+    features: clap_cargo::Features,
+
+    // Top-level flags
+    /// Do not pull in new "audits".
+    #[clap(long)]
+    locked: bool,
+
+    /// How verbose logging should be (log level).
+    #[clap(short, long, arg_enum)]
+    #[clap(default_value_t = Verbose::Warn)]
+    verbose: Verbose,
+
+    /// Instead of stdout, write output to this file.
+    #[clap(long)]
+    output_file: Option<PathBuf>,
+
+    /// Instead of stderr, write logs to this file (only used after successful CLI parsing).
+    #[clap(long)]
+    log_file: Option<PathBuf>,
+}
+
+#[derive(Subcommand)]
+enum Commands {
+    /// initialize cargo-vet for your project
+    Init(InitArgs),
+    /// Fetch the source of `$crate $version`
+    Fetch(FetchArgs),
+    /// Yield a diff against the last reviewed version.
+    Diff(DiffArgs),
+    /// Mark `$crate $version` as reviewed with `$message`
+    Certify(CertifyArgs),
+    /// Mark `$crate $version` as unacceptable with `$message`
+    Forbid(ForbidArgs),
+    /// Suggest some low-hanging fruit to review
+    Suggest(SuggestArgs),
+    /// ??? List audits mechanisms ???
+    Audits(AuditsArgs),
```

Comment:
Need to go back and figure out what this command actually does, I think it's supposed to be kinda a "manager" command like `git remote` or `rustup toolchains`?

Reply:
I don't think it's in the writeup at present.